status-im/open-bounty
# [Fix #259] Migrate to DB backed sessions

Fixes: https://github.com/status-im/open-bounty/issues/259
Status: Finished

## Summary
This migrates from Immutant servlet based sessions to DB backed sessions using [ring.middlware.session](https://ring-clojure.github.io/ring/ring.middleware.session.html), [jdbc-ring-session](https://github.com/luminus-framework/jdbc-ring-session) and [ring-session-timeout](http://ring-clojure.github.io/ring-session-timeout/)

## Why?
- Need to potentially store more data in the session.
- Need to share session information across different application processes.
- A user's session should survive application restarts.

## Things to note:
- A new table `session_store` has been added to hold session information. This will be automatically created when the app starts.
- Any stale session records will be automatically deleted by a session cleaner that runs on the background. A session is considered stale after 6 hours of inactivity.

## QA Notes

(added by @martinklepsch)

- A successful login
- A successful logout
 - After 6 hours of inactivity, they should be logged out and redirected to the home page. The 6 hour period could be extracted to the config and maybe they can change it in their test environment to something shorter (10 min)?
- Session being still present after server restart would be an important thing to test here.


## Alternatives:
I think this might become a bottleneck in the future as every new session hits the DB, but not really sure about this as I don't know the current usage patterns. An alternative could be to use encrypted cookie based sessions, but this depends on the data that's goint to be stored in the session.

I'd appreciate your feedback on the approach as well as on the implementation.

More info:

- http://www.luminusweb.net/docs/sessions_cookies.html
- https://github.com/ring-clojure/ring/wiki/Sessions
